### PR TITLE
Splice with pending committed htlcs

### PR DESCRIPTION
**Pull request by remyers**
This PR updates [splices](https://github.com/ACINQ/eclair/pull/2584) to account for pending htlcs that may exist if the channel has first negotiated [quiescence](https://github.com/ACINQ/eclair/commit/47e0b834383ad7dcdb280c07cbf75bf486e06ee7).

The helper function `makeCommitTxsWithoutHtlcs` is renamed to `makeCommitTxs` and uses the set of committed htlcs that exist when `InteractiveTxBuilder` is instantiated for a splice.

A new attribute `htlcsAmount` is added to `Output.Shared` to account for the value  committed to htlcs from the shared funding input but not assigned to `localAmount` or `remoteAmount`.

### CODE REVIEW

**Comment by remyers** [comment]
I've added modified duplicate versions for some of the tests that already exist in `NormalSplicesStateSpec` to add htlcs updates before the splice testing, and to check the commitment state after the splice. Ultimately I think I should to do one of the following:

 1. integrate these changes into the original tests (eg. add Tag(Quiescence), etc)
 2. move these tests to `NormalQuiescentStateSpec` and focus more on that the commitment state matches what is expected and not include the redundant splice testing logic
 3. move these tests to a new `NormalQuiescentSplicesSpec` and keep the redundant splice testing logic

I'm leaning towards option 1 since we'd like to replace the poor-man's quiescence for all splices. Any preference @t-bast ?

**Comment by t-bast** [comment]
Yes, I think option 1 is best. Each existing tests would become a method, where at the beginning we would have an `if (test.tags.contains(Quiescence))` where you add HTLCs before the splice, and at the end of the test you would have the same `if` to check the end result.

Then each test can be duplicated into two one-liners, something like:

```scala
def testSpliceInCmd(f: FixtureParam, quiescence: Boolean): Unit = {
    ...
}

test("recv CMD_SPLICE (splice-in, no quiescence)") { f =>
    testSpliceInCmd(f, quiescence = false)
}

test("recv CMD_SPLICE (splice-in, quiescence)", Tag(Quiescence)) { f =>
    testSpliceInCmd(f, quiescence = true)
}
```

When we get rid of the poor-man's quiescence, it will make it easy to update.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala [hunk 1]
```diff
@@ -247,9 +251,9 @@ object InteractiveTxBuilder {
     case class Remote(serialId: UInt64, amount: Satoshi, pubkeyScript: ByteVector) extends Output with Incoming
 
     /** The shared output can be added by us or by our peer, depending on who initiated the protocol. */
-    case class Shared(serialId: UInt64, pubkeyScript: ByteVector, localAmount: MilliSatoshi, remoteAmount: MilliSatoshi) extends Output with Incoming with Outgoing {
+    case class Shared(serialId: UInt64, pubkeyScript: ByteVector, localAmount: MilliSatoshi, remoteAmount: MilliSatoshi, htlcsAmount: MilliSatoshi) extends Output with Incoming with Outgoing {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:254]: This seems a bit surprising that we need this here, and that we don't split into local/remote HTLCs...are you sure this is necessary? If so, how come we don't need the same thing for the shared input as well?

I'm wondering if it wouldn't be simpler to include HTLCs inside the `localAmount` and `remoteAmount` (for example by considering that outgoing HTLCs are added to the `remoteAmount` and incoming HTLCs are added to the `localAmount`, as if we're expecting them to be fulfilled), and only when we create the commitment txs at the very end, we recreate the HTLCs by deducing them from these `localAmount` and `remoteAmount` by using the `purpose` field.

I'm not entirely sure, but I think it's worth trying that approach to avoid over-complicating the `Input` and `Output` classes with details that we could keep internal.

Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: I've refactored out he `htlcsAmount` parameter from Output.Shared in [504317d](https://github.com/ACINQ/eclair/pull/2720/commits/504317d2a7542509851ba0d12ddaa40464ee9da1). 

The way I was looking at the `htlcsAmount` was that it was the value that must exist in the new funding tx after local and remote balance changes have been applied. If splice out takes the balance below the amount needed to satisfy the in-flight htlcs then one of the parties could be unable to collect. I also now explicitly check that at the point we make sure the funding output is not below the dust limit.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: After looking into it more, I think storing the `htlcAmount` actually makes a lot of sense, I tried that in https://github.com/ACINQ/eclair/commit/5b7fe763945829b57a1a4ff9e8d0aafdd5075f3b

I still want to add a bit more tests and review the existing ones (in another commit), but we'll be able to compare which approach is simplest once you've finished your investigation.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: I have added more commits to my branch (https://github.com/ACINQ/eclair/tree/splice-with-htlcs) to improve the tests. If including the `htlcAmount` is the right direction forward, you should work on top of my test changes, which highlights that force-close tests weren't interesting because HTLCs were trimmed: we need to verify that we correctly claim the on-chain HTLC txs.

I've also grouped tests together, which makes it easier to see the distinction between the tests that have HTLCs and the pre-quiescence tests. I think the tests themselves can be improved by being more consistent.

Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: In [05ef07d](https://github.com/ACINQ/eclair/pull/2720/commits/05ef07df8177c5e5ae6572f8131c35ad95068b94) I show how we can refactor `Input.Shared` `Output.Shared` to unify local/remote/shared amounts (millisatoshi) into a single amount (satoshi).  

This change primarily impacted:
 - computing `localFees` (used for tests and reporting command success during dual funding)
 - logic of `shouldSignFirst` (I also noticed that this already deviates from initial draft splice spec)
 - `InteractiveTxBuilderSpec` needs further review to make sure local/remote fees are properly checked
 - I am not certain local/remote pushes of value are properly accounted for everywhere; needs more tests

This investigation shows that combining the values for local/remote/shared amounts has minimal downside. After this change we only store these amounts in one place (pre/post splice commitments) and allow the shared inputs/outputs to use whole satoshi amounts.

If we decide to go with this approach, I'll need to integrate the test changes from @t-bast , and generally focus on more tests.

**Comment by remyers** [comment]
We decided to stick with including the local/remote/shared values in the shared inputs/outputs to help with future logging and to make it easier to reason about when looking at the code.

**Review by t-bast — COMMENTED** [review]
The code looks good to me, and it's a really simple change, easy to reason about :+1: 

I'll need to spend a bit more time reviewing the tests: once you're happy with them, can you ping me and move this PR out of draft?

**Comment by remyers** [comment]
[fdaeadd](https://github.com/ACINQ/eclair/pull/2720/commits/fdaeadd8596b17a818eca2156c27d434bfbd0897) unifies the tests with and without quiescence. The force close tests should also check that claim-htlc txs are published.

In the case of `"force-close with multiple splices (simple, quiescence)"` I need to do more research to determine why no 3rd stage htlc claim txs are published. Any ideas @t-bast ?

**Comment by remyers** [comment]
I cleaned up `signCommitTx` and `WaitingForSigs.receiveCommitSig` from `InteractiveTxBuilder` in [81a4814](https://github.com/ACINQ/eclair/pull/2720/commits/81a48146495f74c835890aa8a2e8393d318fc385) to reduce the amount of copy/paste logic. This is especially important to ensure `receiveCommitSig` validates received signatures in a well tested way.

@t-bast  I wasn't entirely happy with my solution to add a new `isSplice: Boolean = false` parameter to the existing `receiveCommit`. Is there a cleaner way to handle this? Is it correct to not increment the index of the local commit when signing for a splice? This is an area where I might need to add more tests to better understand it.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala [hunk 2]
```diff
@@ -615,9 +615,10 @@ case class Commitment(fundingTxIndex: Long,
     // ourCommit.index + 2 -> which is about to become our next revocation hash
     // we will reply to this sig with our old revocation hash preimage (at index) and our next revocation hash (at index + 1)
     // and will increment our index
+    val localCommitIndex = if (isSplice) localCommit.index else localCommit.index + 1
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala:618]: This is a very leaky abstraction, we can do better. Since splices don't actually have a `Commitment` object yet on which to call `receiveCommit`, it makes more sense to move most of the contents of `receiveCommit` to an outside helper function.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala [hunk 3]
```diff
@@ -734,11 +741,16 @@ private class InteractiveTxBuilder(replyTo: ActorRef[InteractiveTxBuilder.Respon
...
+      case Right((localSpec, localCommitTx, remoteSpec, remoteCommitTx, sortedHtlcTxs)) =>
         require(fundingTx.txOut(fundingOutputIndex).publicKeyScript == localCommitTx.input.txOut.publicKeyScript, "pubkey script mismatch!")
         val fundingPubKey = keyManager.fundingPublicKey(channelParams.localParams.fundingKeyPath, purpose.fundingTxIndex)
         val localSigOfRemoteTx = keyManager.sign(remoteCommitTx, fundingPubKey, TxOwner.Remote, channelParams.channelFeatures.commitmentFormat)
-        val localCommitSig = CommitSig(fundingParams.channelId, localSigOfRemoteTx, Nil)
+        val localPerCommitmentPoint = keyManager.htlcPoint(keyManager.keyPath(channelParams.localParams, channelParams.channelConfig))
+        val htlcSignatures = fundingParams.sharedInput_opt match {
+          case Some(_: Multisig2of2Input) => sortedHtlcTxs.map(keyManager.sign(_, localPerCommitmentPoint, purpose.remotePerCommitmentPoint, TxOwner.Remote, channelParams.commitmentFormat)).toList
+          case _ => Nil
+        }
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:752]: The pattern matching is unnecessary, when it's not a splice, `sortedHtlcTxs` will be empty.

**Review by t-bast — APPROVED** [review]

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala (hunk 2 above)
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala]: Thanks - this is a much cleaner solution. Pulled in [08d7f35](https://github.com/ACINQ/eclair/pull/2720/commits/08d7f352e501d46bed498d6ed468ab838972f2db) unchanged.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala (hunk 3 above)
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: Pulled in [08d7f35](https://github.com/ACINQ/eclair/pull/2720/commits/08d7f352e501d46bed498d6ed468ab838972f2db) unchanged.

File: eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala [hunk 4]
```diff
@@ -1081,9 +1237,32 @@ class NormalSplicesStateSpec extends TestKitBaseClass with FixtureAnyFunSuiteLik
...
     alice ! WatchFundingSpentTriggered(commitTx2)
+
+    // watch for all htlc outputs from local commit-tx to be spent
+    val watchHtlcsOut = htlcs.aliceToBob.map(_ => alice2blockchain.expectMsgType[WatchOutputSpent])
+    htlcs.bobToAlice.map(_ => alice2blockchain.expectMsgType[WatchOutputSpent])
+    // trigger spent, watch for confirmation and confirm alice's htlc-timeout txs
+    (watchHtlcsOut zip htlcsTxsOut).foreach { case (watch, tx) => watch.replyTo ! WatchOutputSpentTriggered(tx) }
+    htlcsTxsOut.foreach { tx =>
+      assert(tx.txid == alice2blockchain.expectMsgType[WatchTxConfirmed].txId)
+      alice ! WatchTxConfirmedTriggered(BlockHeight(400000), 42, tx)
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala]: Improvements look good in [15bdebf](https://github.com/ACINQ/eclair/pull/2720/commits/15bdebf39cd024d164b11a1aecc464db1cdff643). Good idea to add a test for bad/out-of-order signatures too.

**Comment by t-bast** [comment]
LGTM, let's just wait for #2613 to get in first, and then we can rebase and merge :+1:

File: eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala [hunk 5]
```diff
@@ -986,8 +1143,19 @@ class NormalSplicesStateSpec extends TestKitBaseClass with FixtureAnyFunSuiteLik
     bob2alice.forward(alice)
     awaitCond(alice.stateData.asInstanceOf[DATA_NORMAL].commitments.active.size == 1)
     awaitCond(bob.stateData.asInstanceOf[DATA_NORMAL].commitments.active.size == 1)
+
+    resolveHtlcs(f, htlcs, spliceOutFee = 0.sat)
   }
 
+  test("disconnect (tx_signatures received by alice, zero-conf)", Tag(ChannelStateTestsTags.ZeroConf), Tag(ChannelStateTestsTags.AnchorOutputsZeroFeeHtlcTxs)) { f =>
```
Comment (remyers) [inline eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala:1150]: Squashed and rebased in commit [ee8d325](https://github.com/ACINQ/eclair/commit/ee8d3253ccb9b708c0dec26f541de7563c8da9ae) on latest master.

The merge included adding a quiescence version for the test `"disconnect (tx_signatures received by alice, zero-conf)"`.

**Review by t-bast — APPROVED** [review]
LGTM, good job!
